tonic/src/transport/server/tls.rs

```diff
@@ -38,15 +38,29 @@ impl ServerTlsConfig {
         }
     }
 
-    /// Sets a certificate against which to validate client TLS certificates.
-    pub fn client_ca_root(self, cert: Certificate) -> Self {
+    /// Sets the [`ClientTlsAuth`] approach.
+    pub fn client_auth(self, client_auth: ClientTlsAuth) -> Self {
         ServerTlsConfig {
-            client_ca_root: Some(cert),
+            client_auth,
             ..self
         }
     }
 
     pub(crate) fn tls_acceptor(&self) -> Result<TlsAcceptor, crate::Error> {
-        TlsAcceptor::new(self.identity.clone().unwrap(), self.client_ca_root.clone())
+        TlsAcceptor::new(self.identity.clone().unwrap(), self.client_auth.clone())
     }
 }
+
+/// Client TLS authentication options.
+#[cfg(feature = "tls")]
+#[cfg_attr(docsrs, doc(cfg(feature = "tls")))]
+#[derive(Clone, Debug, Default)]
+pub enum ClientTlsAuth {
+    /// No client authentication.
+    #[default]
+    None,
+    /// Optional client authentication with a provided root CA certificate.
+    Optional(Certificate),
```

Comment:
Could you explain more what the optional use case is? Do you know if any of the other gRPC implementations provide this sort of feature? As a side note, if this is a more niche use case you may want to consider using hyper directly and using `hyper-rustls` which will give you full access to all the rustls settings.

Reply:
This option is useful, for example, when you have two types of clients accessing your service. One type may be only viewers that can access some data you provide through your endpoints, and the other may have write access but have to authenticate first. Having to issue certificates to all viewers who may access your endpoints only once may not be ideal. Furthermore, you cannot issue such certificates over the gRPC with mandatory authentication (because the clients would first need to have the certificate).

The `Optional(Certificate)` option guarantees that if a client included a certificate, it was validated by standard procedures, but it may choose not to include any. This way, it is much safer than avoiding certificate verification altogether and doing it by yourself later on for endpoints that should require authentication.

I have not used any other gRPC libraries prior to `tonic`, but a quick search for a gRPC library led me to the grpc.io C++ library and to this page https://grpc.github.io/grpc/cpp/grpc__security__constants_8h.html#a29ffe63a8bb3b4945ecab42d82758f09 where are listed five different client authentication options. The enum values that I have included in the pull request correspond to the constants this way:
```
None <=> GRPC_SSL_DONT_REQUEST_CLIENT_CERTIFICATE 	
Optional(Certificate) <=> GRPC_SSL_REQUEST_CLIENT_CERTIFICATE_AND_VERIFY
Mandatory(Certificate) <=> GRPC_SSL_REQUEST_AND_REQUIRE_CLIENT_CERTIFICATE_AND_VERIFY
```

The remaining two options may be useful when you intend to perform certificate verification by yourself, but that can (and should) be considered dangerous. I do not currently see any good use cases for these options, so they are not included in  this pull request. They can be added by extending the enum that I proposed by additional fields that would take a trait object implementing `rustls::server::ClientCertVerifier`, but I would consider this quite a controversial addition that would be better left to using `hyper` directly if needed, as you mentioned.

Reply:
Yeah, that seems fair!